Some performance tweaks

Ptico opened the pull request:
As library used by another libraries, inflecto should work as fast as possible. Here is some performance tweaks i planned during previous PR. Results on ruby 2.1.3p242:

* tableize - (1.3-1.6x)
* ordinalize - (1.1-1.3x)
* uncountable? - (2x)

pluralize and singularize should work faster as well, because they use uncountable

Also this PR demonstrates my crazy naming skills, so feel free to suggest better names

mbj commented on spec/unit/inflecto/inflections/plural_spec.rb:
```diff
@@ -24,7 +24,7 @@
 
     it 'removes it from uncountables' do
-      subject.uncountables.should eq([])
+      subject.uncountables.should be_empty
```
I personally think the `eq([])` matcher is more exact than `be_empty` I realize this change is from transforming `uncountables` into a `Set` instance. So we IMO should match against `eql(Set.new)`.

Note that I prefer `eql` over `eq` as the former translates to more strict `Object#eql?` where the latter is `Object#==` that does coercion in some cases.

Ptico replied:
I also prefer equal > eql > eq strategy, but usually for collections i use less restrictive matchers to have less reason to change specs. But you are right: in this case eql is better

mbj commented on spec/unit/inflecto/inflections/plural_spec.rb:
```diff
@@ -32,7 +32,7 @@
 
     it 'removes it from uncountables' do
-      subject.uncountables.should eq([])
+      subject.uncountables.should be_empty
```
See https://github.com/mbj/inflecto/pull/8/files#r22766596

mbj commented on spec/unit/inflecto/inflections/singular_spec.rb:
```diff
@@ -24,7 +24,7 @@
 
     it 'removes it from uncountables' do
-      subject.uncountables.should eq([])
+      subject.uncountables.should be_empty
```
See https://github.com/mbj/inflecto/pull/8/files#r22766596

mbj commented on spec/unit/inflecto/inflections/singular_spec.rb:
```diff
@@ -32,7 +32,7 @@
 
     it 'removes it from uncountables' do
-      subject.uncountables.should eq([])
+      subject.uncountables.should be_empty
```
See https://github.com/mbj/inflecto/pull/8/files#r22766596

mbj commented on spec/unit/inflecto/inflections/uncountable_spec.rb:
```diff
@@ -5,16 +5,16 @@
   context 'when word given' do
     subject { object.uncountable('water') }
-    its(:uncountables) { should eq(['water']) }
+    its(:uncountables) { should contain_exactly('water') }
```
I'd prefer to see `should eql(Set.new(%w[water]))` here for reasons in https://github.com/mbj/inflecto/pull/8/files#r22766596.

mbj commented on spec/unit/inflecto/inflections/uncountable_spec.rb:
```diff
@@ -5,16 +5,16 @@
   context 'when array of words given' do
     subject { object.uncountable(['water', 'sugar']) }
-    its(:uncountables) { should eq(%w[water sugar]) }
+    its(:uncountables) { should contain_exactly('water', 'sugar') }
```
See https://github.com/mbj/inflecto/pull/8/files#r22766606

mbj commented on spec/unit/inflecto/inflections/uncountable_spec.rb:
```diff
@@ -5,16 +5,16 @@
   context 'when multiple words given' do
     subject { object.uncountable('water', 'sugar') }
-    its(:uncountables) { should eq(%w[water sugar]) }
+    its(:uncountables) { should contain_exactly('water', 'sugar') }
```
See https://github.com/mbj/inflecto/pull/8/files#r22766606

mbj commented on spec/unit/inflecto/class_methods/uncountable_predicate_spec.rb:
```diff
@@ -0,0 +1,19 @@
+
+describe Inflecto, '.uncountable?' do
+  it 'truthy when empty string' do
+    Inflecto.uncountable?('').should == true
```
We should use the most strict matcher here. `Object#==` is more permissive than `Object#eql?` than `Object#equal?`. I think `should be(true)` is the strictest and less lowerfullest primitive to use here.

mbj commented on spec/unit/inflecto/class_methods/uncountable_predicate_spec.rb:
```diff
@@ -0,0 +1,19 @@
+  end
+
+  it 'truthy when word is present in list' do
+    Inflecto.uncountable?('grass').should == true
```
See https://github.com/mbj/inflecto/pull/8/files#r22766616

mbj commented on spec/unit/inflecto/class_methods/uncountable_predicate_spec.rb:
```diff
@@ -0,0 +1,19 @@
+  end
+
+  it 'falsy when word is not present in list' do
+    Inflecto.uncountable?('user').should == false
```
See https://github.com/mbj/inflecto/pull/8/files#r22766616

mbj commented on spec/unit/inflecto/class_methods/uncountable_predicate_spec.rb:
```diff
@@ -0,0 +1,19 @@
+  end
+
+  it 'truthy when word is present in list but in different case' do
+    Inflecto.uncountable?('FiSH').should == true
```
See https://github.com/mbj/inflecto/pull/8/files#r22766616

Also `FiSH` might be a typo of `Fish` ?

Ptico replied:
Yes, but i leave it to make sure that it works with different-cased words

mbj commented on lib/inflecto/defaults.rb:
```diff
@@ -54,5 +54,5 @@
   inflect.irregular('cow', 'cows')
 
-  inflect.uncountable(%w(hovercraft moose milk rain Swiss grass equipment information rice money species series fish sheep jeans))
+  inflect.uncountable(%w(hovercraft moose milk rain swiss grass equipment information rice money species series fish sheep jeans))
```
I think the lowercase word `swiss` does not exist in english, where the capitalized `Swiss` exists as a name that is always capitalized. We should keep capitalization in this case.

mbj commented:
@Ptico I did a first pass over the code. I'm curious of my reasoning makes sense to you.

I'm not a native speaker, and for that reason should NOT maintain this library. Also I do not have any project that uses it (anymore). This results in an overall lack of attention from my side.

I hope we can bring this lib into a "stable state" where all mutations are dead and the API is kinda stable so we can look for someone to maintain it in future.

Ptico commented:
@mbj i will fix your comments soon. I agree that maintainer should be a native speaker and will be happy to help make it stable. I like inflecto and similar small support libs, because they fix all what i hate in AS. So it worth to continue its development

Ptico commented:
The only question: may i exclude tests from flay scoring in this PR? IMO specs is for reading and i don't think its reasonable to DRY them.

mbj commented:
@Ptico Flay is just an indicator, when the score moves its the programmers decision if the move is appropriate or note. Feel free to simply set the flay score to whatever value it needs to have to be green.

semmons99 commented:
@Ptico any movement on this? Can I help you?

Ptico commented:
Hi! I'll update flay score in an hour and it's ready

Ptico commented:
I have a mess with devtools again. @semmons99 may you try to clone my branch and go around rake metrics:* tasks? I can't figure out why metrics:mutant and metrics:flay stopped working.

Also, looks like rubocop rules have been changed, @mbj are you ok with changes like double quotes to single quotes etc.?

mbj commented:
@Ptico Sure, feel free to quote in a way the least powerful primitive that does the job (in this case the single quotes) is in use.

Devtools was recently changed not to require the `Gemfile.devtools` anymore. Just nuke it.

semmons99 commented:
I feel like a fool. I can't even get `bundle install` to run right now.

```
λ bundle install
Your Gemfile lists the gem inflecto (>= 0) more than once.
You should probably keep only one of them.
While it's not a problem now, it could cause errors if you change the version of just one of them later.
Updating https://github.com/rom-rb/devtools.git
Fetching gem metadata from https://rubygems.org/.........
Resolving dependencies...
Bundler could not find compatible versions for gem "rake":
  In Gemfile:
    devtools (>= 0) ruby depends on
      rake (~> 10.4.0) ruby

    rake (10.3.2)
```

Ptico commented:
Hi @mbj i'm back to inflecto :)
I've fixed devtools, but flog rake task can't be passed because float threshold is not allowed (undefined method `succ' for 11.3:Float)

Also, because minimum required version is 1.9.3 and 1.8 reached EOL i removed 1.8 specific code

mbj commented:
@Ptico Use a `Fixnum` instead a `Float` literal in the `config/flay.yml`. It should actually check that, are you sure you use very latest devtools?

> Also, because minimum required version is 1.9.3 and 1.8 reached EOL i removed 1.8 specific code

+1

Ptico commented:
Hm, if i use 12 it doesn't run next checks. devtools v 0.1.2. Will investigate a bit deeper

Ptico commented:
@mbj now, i'm stuck with mutant. I have some mutants like 

```diff
-  @__instance__ ||= new
+  @__instance__ ||= self
```

and the coverage is 99.7, but i can't set proper `expect_coverage` because non of the 99.7, 99, 0.99 and 0.997 works. I receive non-zero exit status anyway.

mbj commented:
> @mbj now, i'm stuck with mutant. I have some mutants like

Kill that one via asserting object identity, not equvalency. (if identity is correct here, else go with the mutation).

Ptico commented:
Hi @mbj ! I still have no luck with mutant, but i found the way to set expected coverage (which is higher than before btw). Let's merge this PR already, it improves a lot and almost 2 years old

mbj commented:
@Ptico Yes this PR is an improvement.

Ptico commented:
Yay! I'll open a bottle of champagne for this! :)
